File: elixir/perfect-numbers/lib/perfect_numbers.ex
defmodule PerfectNumbers do
  @doc """
  Determine the aliquot sum of the given `number`, by summing all the factors
  of `number`, aside from `number` itself.

  Based on this sum, classify the number as:

  :perfect if the aliquot sum is equal to `number`
  :abundant if the aliquot sum is greater than `number`
  :deficient if the aliquot sum is less than `number`
  """
  @spec classify(number :: integer) :: {:ok, atom} | {:error, String.t()}
  def classify(number) when not is_integer(number) or number <= 0,
    do: error("Classification is only possible for natural numbers.")

  def classify(number) do
    aliquote_sum(number)
    |> type(number)
    |> ok()
  end

  defp type(_, 1), do: :deficient
  defp type(number, number), do: :perfect
  defp type(sum, number) when sum > number, do: :abundant
  defp type(sum, number) when sum < number, do: :deficient

  defp aliquote_sum(number) do
    1..ceil(number / 2)
    |> Enum.filter(fn x -> rem(number, x) == 0 end)
    |> Enum.reduce(0, fn x, sum -> sum + x end)
  end

  defp ok(value) do
    {:ok, value}
  end

  defp error(reason) do
    {:error, reason}
  end
end

File: javascript/perfect-numbers/perfect-numbers.js
//@ts-check
const NicomachusCategory = Object.freeze({
  PERFECT: "perfect",
  ABUDANT: "abundant",
  DEFICIENT: "deficient"
});

/**
 * 
 * @param {number} x 
 * @returns {boolean}
 */
function isNatural(x) {
  return Number.isInteger(x) && x > 0;
}

/**
 * 
 * @param {number} value 
 */
function aliquote(value) {
  const aliquote = [];

  for (let integer = 1; integer <= value/2; integer++) {
    const div = value / integer;

    if(isNatural(div)) {
      aliquote.push(integer);
    }

  }

  return aliquote;

}

/**
 * 
 * @param {number} value 
 * @returns {string}
 */
export const classify = (value) => {

  if (!isNatural(value)) {
    throw new Error("Classification is only possible for natural numbers.");
  }

  const aliquoteSum = aliquote(value)
    .reduce((acc, value) => acc + value, 0);

  if (aliquoteSum == value) {
    return NicomachusCategory.PERFECT;
  }

  if (aliquoteSum > value) {
    return NicomachusCategory.ABUDANT;
  }

  return NicomachusCategory.DEFICIENT;

};
